### PR TITLE
Remove parameters with `deparam`

### PR DESCRIPTION
**Pull request by dlecocq**
@b4hand @tanglyh @tammybailey @martin-seomoz

### CODE REVIEW

File: src/url.cpp [hunk 1]
```diff
@@ -457,4 +457,54 @@ namespace Url
...
+    Url& Url::deparam(const std::unordered_set<std::string>& blacklist)
+    {
+        remove_params(params_, blacklist, ';');
+        remove_params(query_, blacklist, '&');
+        return *this;
+    }
+
+    void Url::remove_params(std::string& str, const std::unordered_set<std::string>& blacklist, const char separator)
+    {
+
```
Comment (b4hand) [inline src/url.cpp]: extra blank line?

File: src/url.cpp [hunk 2]
```diff
@@ -457,4 +457,54 @@ namespace Url
...
+            {
+                pieces.push_back(piece);
+            }
+        }
+
+        if (previous < str.length())
+        {
+            std::string piece = str.substr(previous);
+            std::string name = piece.substr(0, piece.find('='));
+            std::transform(name.begin(), name.end(), name.begin(), ::tolower);
```
Comment (b4hand) [inline src/url.cpp:489]: I'm pretty sure that `::tolower` will only do the right thing for the ASCII range. Isn't this an issue if you have UTF-8 encoded strings? Furthermore, this is probably an issue throughout where you are doing "case-insensitive" comparisons.

File: src/url.cpp (hunk 1 above)
Comment (dlecocq) [inline src/url.cpp]: Removed

File: src/url.cpp (hunk 2 above)
Comment (dlecocq) [inline src/url.cpp:489]: That's my understanding, too, and I suspect that it's the right / acceptable thing to do in this case. Any parameters we want to remove that aren't in ASCII will have to be escaped anyway, so it's not clear what to do in that case anyway -- when we want to replace `bücher` do we replace `b%C3%9Ccher` (bÜcher) _and_ `b%C3%BCcher` (bücher)? Or would we expect the argument to `deparam` to be the encoded version?

And then WRT hostnames (another spot where we use `::tolower`), I'm earnestly not sure if `bücher.com` and `bÜcher.com` are the same name -- they definitely punycode differently.

Thoughts?

Comment (dlecocq) [inline src/url.cpp:489]: In-person discussion revealed that:

1) we're not going to support (at least for now) case-insensitivity for non-ASCII params
2) we don't exactly know what to do with hostnames -- it seems that https://www.ietf.org/rfc/rfc3454.txt says one thing (and mozscape does it -- https://github.com/seomoz/moz-util-py/blob/master/test/test_convert_url.py#L74-L79), but http://unicode.org/faq/idn.html#11 says another. We'll make a ticket to follow up about that.

**Pull request merged**